src/ray/core_worker/transport/out_of_order_actor_scheduling_queue.cc

```diff
@@ -0,0 +1,122 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/transport/out_of_order_actor_scheduling_queue.h"
+
+namespace ray {
+namespace core {
+
+OutOfOrderActorSchedulingQueue::OutOfOrderActorSchedulingQueue(
+    instrumented_io_context &main_io_service, DependencyWaiter &waiter,
+    std::shared_ptr<PoolManager> pool_manager, bool is_asyncio, int fiber_max_concurrency,
+    const std::vector<ConcurrencyGroup> &concurrency_groups)
+    : main_thread_id_(boost::this_thread::get_id()),
+      waiter_(waiter),
+      pool_manager_(pool_manager),
+      is_asyncio_(is_asyncio) {
+  if (is_asyncio_) {
+    std::stringstream ss;
+    ss << "Setting actor as asyncio with max_concurrency=" << fiber_max_concurrency
+       << ", and defined concurrency groups are:" << std::endl;
+    for (const auto &concurrency_group : concurrency_groups) {
+      ss << "\t" << concurrency_group.name << " : " << concurrency_group.max_concurrency;
+    }
+    RAY_LOG(INFO) << ss.str();
+    fiber_state_manager_ =
+        std::make_unique<FiberStateManager>(concurrency_groups, fiber_max_concurrency);
+  }
+}
+
+void OutOfOrderActorSchedulingQueue::Stop() {
+  if (pool_manager_) {
+    pool_manager_->Stop();
+  }
+}
+
+bool OutOfOrderActorSchedulingQueue::TaskQueueEmpty() const {
+  RAY_CHECK(false) << "TaskQueueEmpty() not implemented for actor queues";
+  return false;
+}
+
+size_t OutOfOrderActorSchedulingQueue::Size() const {
+  RAY_CHECK(false) << "Size() not implemented for actor queues";
+  return 0;
+}
+
+void OutOfOrderActorSchedulingQueue::Add(
+    int64_t seq_no, int64_t client_processed_up_to,
+    std::function<void(rpc::SendReplyCallback)> accept_request,
+    std::function<void(rpc::SendReplyCallback)> reject_request,
+    rpc::SendReplyCallback send_reply_callback, const std::string &concurrency_group_name,
+    const ray::FunctionDescriptor &function_descriptor,
+    std::function<void(rpc::SendReplyCallback)> steal_request, TaskID task_id,
+    const std::vector<rpc::ObjectReference> &dependencies) {
+  // A seq_no of -1 means no ordering constraint.
+  RAY_CHECK(boost::this_thread::get_id() == main_thread_id_);
+
+  auto request = InboundRequest(std::move(accept_request), std::move(reject_request),
+                                std::move(steal_request), std::move(send_reply_callback),
+                                task_id, dependencies.size() > 0, concurrency_group_name,
+                                function_descriptor);
+
+  if (dependencies.size() > 0) {
+    waiter_.Wait(dependencies, [this, request = std::move(request)]() mutable {
```

Comment:
i think this might leak the request if the dependency is not resolved forever. Maybe we should improve this?

Reply:
cc @rkooo567

Reply:
I think we're guaranteed dependencies will be resolved eventually...

Reply:
@ericl Is there any good doc that explains the guarantee? (maybe whitepaper?) Maybe we can comment the link here..? I can imagine if the upstream task is something like while True, it won't be resolved forever, but I guess that's the application bug that it has such code.

Reply:
It's just part of the Ray API guarantee... if you .remote(x) then we wait for x to be resolved. The underlying protocol is reliable and guarantees we eventually detect if x is failed or completed.